Lazy pulse qobj loading for large backend

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Loading backend `PulseDefaults` with 100+ qubit is significantly slow because it immediately converts all pulse qobjs into `Schedule`s when it is called for the first time. This induces significant performance regression in transpile because all preset passes require the default object, i.e. `InstructionScheduleMap`, to invoke pulse gate pass for user calibration.

This PR allows us to convert it as-needed basis for speed up.

Fix https://github.com/Qiskit/qiskit-terra/issues/7914

Blocked by https://github.com/Qiskit/qiskit-terra/pull/8839

### Details and comments

This PR allows `InstructionScheduleMap` to take unconverted pulse qobj, and invoke the conversion logic when the `InstructionScheduleMap.get` method is called for a particular calibration. 

This is mainly done by the commit 64a37122e64d7cb98b75cc79a8574759460592d0

![image](https://user-images.githubusercontent.com/39517270/195410961-6e6f383a-74d0-4a07-9f3f-f241fbdca879.png)

According to the profile, I also removed the use of deep copy (this is added because Qobj data is a nested dict and to avoid mutating it) in the commit 0e1b5f2ef5756e5ba625e21e0c95d5372a86af02 (roughly 3x speedup)

![image](https://user-images.githubusercontent.com/39517270/195411235-05dfa39d-52bc-4820-890e-2b14679b006f.png)

This is output of airspeed velocity test:

```
Benchmarks that have improved:

       before           after         ratio
     [e9736b41]       [0e1b5f2e]
     <fix/pulse_qobj_converter>       <upgrade/lazy_instmap_qobj_conversion>
-        774±10ms       21.0±0.5ms     0.03  pulse.load_pulse_defaults.PulseDefaultsBench.time_building_defaults(1000)
-        88.1±2ms      2.09±0.08ms     0.02  pulse.load_pulse_defaults.PulseDefaultsBench.time_building_defaults(100)
-      14.8±0.6ms          292±6μs     0.02  pulse.load_pulse_defaults.PulseDefaultsBench.time_building_defaults(10)
-      6.66±0.3ms          112±3μs     0.02  pulse.load_pulse_defaults.PulseDefaultsBench.time_building_defaults(0)

Benchmarks that have stayed the same:

       before           after         ratio
     [e9736b41]       [0e1b5f2e]
     <fix/pulse_qobj_converter>       <upgrade/lazy_instmap_qobj_conversion>
         17.6±1ms       18.4±0.6ms     1.05  pulse.load_pulse_defaults.CircuitSchedulingBench.time_scheduling_circuits(3)
         74.8±2ms         73.0±3ms     0.98  pulse.load_pulse_defaults.CircuitSchedulingBench.time_scheduling_circuits(15)

Benchmarks that have got worse:

       before           after         ratio
     [e9736b41]       [0e1b5f2e]
     <fix/pulse_qobj_converter>       <upgrade/lazy_instmap_qobj_conversion>
+      6.74±0.3ms       9.98±0.3ms     1.48  pulse.load_pulse_defaults.CircuitSchedulingBench.time_scheduling_circuits(1)
+      12.6±0.7ms         16.0±2ms     1.27  pulse.load_pulse_defaults.CircuitSchedulingBench.time_scheduling_circuits(2)
```


As you can see we achieved roughly x50 speedup with these two commits. Note that scheduling test suffers the performance regression as a side effect of the lazy conversion (because it internally `.get` schedule). I guess the same problem will occur in the V2 converter https://github.com/Qiskit/qiskit-terra/pull/8759, but the converter should be able to avoid the performance regression by implementing the similar lazy conversion mechanism. Regarding circuit scheduler, in principle, we should improve the mechanism to build `Schedule`, as proposed in https://github.com/Qiskit/qiskit-terra/pull/8029.

mtreinish commented:
> I guess the same problem will occur in the V2 converter https://github.com/Qiskit/qiskit-terra/pull/8759, but the converter should be able to avoid the performance regression by implementing the similar lazy conversion mechanism.

All the IBM provider (except for the one with the q) backends will have this same problem too since they all pull in the pulse schedules from defaults for `InstructionProperties` during the `Target` generation

nkanazawa1989 commented:
> All the IBM provider (except for the one with the q) backends will have this same problem too since they all pull in the pulse schedules from defaults for InstructionProperties during the Target generation

Do they have own mechanism for Target construction or just using some code in terra? If they rely on terra code I can update InstructionProperties in follow-up/in this PR.

mtreinish commented:
It's a copy and paste of the same basic code that's in the `BackendV2Converter`, which was all based on a rough draft that I wrote about ~8 months ago for taking the IBM JSON response payloads and creating a `Target` from the dictionaries :

- https://github.com/Qiskit/qiskit-ibm-provider/blob/main/qiskit_ibm_provider/utils/backend_converter.py
- https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/utils/backend_converter.py

nkanazawa1989 commented:
Added new commit https://github.com/Qiskit/qiskit-terra/pull/8885/commits/26903089dcf10120e3f1c8b5e0accdcf9c8256d4 to support lazy conversion for V2. This commit allows `InstructionProperties` to take `PulseQobjDef` instance in addition to `Schedule` and `ScheduleBlock`. Then the conversion logic is invoked when `InstructionProperties.calibration` is called.

mtreinish commented on qiskit/providers/models/pulsedefaults.py:
```diff
@@ -108,7 +106,7 @@ def __init__(self, name: str, qubits=None, sequence=None, **kwargs):
             name (str): The name of the command
             qubits: The qubits for the command
-            sequence (PulseQobjInstruction): The sequence for the Command
+            sequence (List[PulseQobjInstruction]): The sequence for the Command
```
I assume this was just a doc bug based on the argument name being sequence and not an api change. So this always took a list and the type hint was just wrong.

nkanazawa1989 replied:
Okey, I reverted the change in [8c82a02](https://github.com/Qiskit/qiskit-terra/pull/8885/commits/8c82a021c23a2002670d56c24f9d8e03e39db3e5)

mtreinish reviewed (commented):
I'm a big fan of this, it'll fix most of the overhead for circuits that aren't using pulse and avoid the need to parse the entire payload which can be super expensive. Thanks for doing this! Looking through the code as a first pass nothing major stands out to me, but I'll do a more through pass after #8839 merges and this is rebased. 

The one thing I think will be good to have validation on is compilation across a multiprocessing boundary works as expected. I don't see anything super concerning on that front, but since we're moving towards lazy loading the defaults having tests that will do that across a process boundary seems like a good thing to add. We have this class:

https://github.com/Qiskit/qiskit-terra/blob/main/test/python/compiler/test_transpiler.py#L1817

If you adds some tests which require using the pulse calibrations from the defaults file for something in `transpile()` with multiple circuits along with `PassManager.run()` with multiple circuits too they have different parallel dispatch methods (although the current `transpile()` one will go away in 0.25.0 and just use `PassManager.run()` internally).

nkanazawa1989 commented:
Thanks Matthew. I rebased the branch and added a new test file for calibration entry classes and updated the transpile test with the lazy loading test case. Because we don't have any preset pass that uses the backend calibration (we have RZX calibration builder but this is not in the preset passes), I wrote a fake pass for testing.

mtreinish commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -321,12 +283,14 @@ def remove(
-        self._qubit_instructions[qubits].remove(instruction)
+
+        del self._map[instruction][qubits]
         if not self._map[instruction]:
-            self._map.pop(instruction)
+            del self._map[instruction]
+
+        self._qubit_instructions[qubits].remove(instruction)
         if not self._qubit_instructions[qubits]:
-            self._qubit_instructions.pop(qubits)
+            del self._qubit_instructions[qubits]
```
This seems like a bit of an unrelated change, it's better to keep the diff limited to the logical change of the PR. The logic looks the same to me so it's not anything worth blocking over.

nkanazawa1989 replied:
Yeah strictly speaking this is unrelated to this change but I realized this logic is slower according to this [stackoverflow](https://stackoverflow.com/questions/24547223/is-del-or-pop-preferred-when-removing-elements-from-dicts). So I made this change to improve overall performance of inst map.

mtreinish reviewed (commented):
Overall this LGTM, I like the new structure and the definition of the interface for a calibration entry in a instruction schedule map is a good addition. Thanks for adding the additional test coverage with the parallel context, I didn't expect any issue with it, but think it's good to have.

I had a question inline and a spot where it looks like an unrelated change (not blocking). The only thing missing is maybe some release notes. We're changing a bit of the `InstructionScheduleMap` interface particularly around types supported. You might also want a feature note for the lazy loading, because the behavior will be a bit different (although equivalent) and provide a nice speedup which would good to talk about in the release documentation (also maybe talking about the new calibration entry interface, although that might not really be intended to be public, so we can skip it).

nkanazawa1989 commented on qiskit/providers/fake_provider/utils/backend_converter.py:
```diff
@@ -113,17 +113,19 @@ def convert_to_target(conf_dict: dict, props_dict: dict = None, defs_dict: dict
+                    qargs = (qarg,)
+                # Do NOT call .get method. This parses Qpbj immediately.
+                # This operation is computationally expensive and should be bypassed.
+                calibration_entry = inst_map._map[inst][qargs]
```
Sounds like a good idea. Done in [add API for getting calibration entry](https://github.com/Qiskit/qiskit-terra/pull/8885/commits/41217fdfaf087579e9d92efc16d7dde654f81e34). I added new API as a protected method since this is not expected to interface with users. During this change I also noticed backend converter exists for different places, namely `qiskit.providers.fake_provider.utils.backend_converter.convert_to_target` and `qiskit.providers.backend_compat.convert_to_target`. Do you know why the similar code exists in different places.

mtreinish replied:
I think because in #8759 what I wrote for that PR was more general and I intended to follow up and remove the fake provider version which was more targeted to ibm backends but forgot to. I can push a PR to do this now

mtreinish replied:
Oh, I guess one key difference between the implementations is the fake provider version works with a raw dictionary without creating an intermediate `BackendConfiguration` and `BackendProperties`  object. IIRC, we did that for efficiency because we can just parse the raw json files and go straight to a target. I'm not sure if it's worth deleting the second copy or not (or how much of a difference that makes in practice). Also at some far future date when `BackendConfiguration` and `BackendProperties` are deprecated and then removed we will want the dictionary version too.

mtreinish reviewed (approved):
LGTM, thanks for the fast update.

mtreinish commented:
Oops before we tag as atuomerge did you want to add a release note?

nkanazawa1989 commented:
Thanks for catching the missing release note. I had forgotten to write the note. Now it's added.
